### PR TITLE
iio: Add support of IIO sampling frequency to channel level

### PR DESCRIPTION
**Pull request by yongli3**
Fix issue #2101

Signed-off-by: Yong Li <yong.b.li@intel.com>

### CODE REVIEW

**Comment by yongli3** [comment]
@laykuanloon @edersondisouza  please help to review/test, I am using the HMC5843 to test it

File: src/lib/io/include/sol-iio.h [hunk 1]
```diff
@@ -72,6 +72,7 @@ typedef struct sol_iio_config {
 #define SOL_IIO_CONFIG_API_VERSION (1)
     uint16_t api_version; /**< The API version */
 #endif
+    char sampling_frequency_name[PATH_MAX];
```
Comment (cabelitos) [inline src/lib/io/include/sol-iio.h]: No no, add this at end of the struct!
Otherwise we will break things.

Comment (yongli3) [inline src/lib/io/include/sol-iio.h]: Done, Thanks!

File: src/lib/io/include/sol-iio.h [hunk 2]
```diff
@@ -77,6 +77,7 @@ typedef struct sol_iio_config {
     const void *data; /**< User defined data to be sent to sol_iio_reader_cb */
     int buffer_size; /**< The size of reading buffer. 0: use device default; -1: disable buffer and readings will be performed on channel files on sysfs. */
     int sampling_frequency; /**< Device sampling frequency. -1 uses device default */
+    char sampling_frequency_name[PATH_MAX];
```
Comment (laykuanloon) [inline src/lib/io/include/sol-iio.h]: Suggest to add comment like other attribute.
Some drivers exposing its sampling frequency that is shared by channel type. For e.g: in_magn_sampling_frequency, in_accel_sampling_frequency.

File: src/lib/io/sol-iio.c [hunk 3]
```diff
@@ -116,6 +116,8 @@ struct resolve_absolute_path_data {
 #define CHANNEL_SCAN_TYPE_PATH SYSFS_DEVICES_PATH "/iio:device%d/scan_elements/%s_type"
 
 #define SAMPLING_FREQUENCY_DEVICE_PATH SYSFS_DEVICES_PATH "/iio:device%d/sampling_frequency"
+#define SAMPLING_FREQUENCY_DEVICE_PATH2 SYSFS_DEVICES_PATH "/iio:device%d/%ssampling_frequency"
```
Comment (laykuanloon) [inline src/lib/io/sol-iio.c]: Suggest put more meaningful define. 
CHANNEL_SAMPLING_FREQUENCY_DEVICE_PATH

**Comment by laykuanloon** [comment]
I tested is working.

**Comment by yongli3** [comment]
@laykuanloon I updated the PR to follow your comments, I also added the accel/gyro devices support, please review

**Comment by laykuanloon** [comment]
@yongli3 
+1

File: src/lib/io/include/sol-iio.h [hunk 4]
```diff
@@ -77,6 +77,7 @@ typedef struct sol_iio_config {
     const void *data; /**< User defined data to be sent to sol_iio_reader_cb */
```
Comment (edersondisouza) [inline src/lib/io/include/sol-iio.h]: Well, as you are changing a versioned structure, I'd say that `SOL_IIO_CONFIG_API_VERSION` may need to be changed - even if only adding stuff. @barbieri ?

File: src/lib/io/include/sol-iio.h [hunk 5]
```diff
@@ -77,6 +77,7 @@ typedef struct sol_iio_config {
...
     int buffer_size; /**< The size of reading buffer. 0: use device default; -1: disable buffer and readings will be performed on channel files on sysfs. */
     int sampling_frequency; /**< Device sampling frequency. -1 uses device default */
+    char sampling_frequency_name[PATH_MAX]; /**< Sampling frequency sysfs node name. Some drivers expose the sampling frequency that is shared by channel type. Such as in_magn_sampling_frequency, in_accel_sampling_frequency. */
```
Comment (edersondisouza) [inline src/lib/io/include/sol-iio.h]: `PATH_MAX` appears overkill. Knowing that this is only part of name, why not `NAME_MAX`?

Comment (yongli3) [inline src/lib/io/include/sol-iio.h]: Done Thanks @edersondisouza !

File: src/lib/io/include/sol-iio.h (hunk 4 above)
Comment (edersondisouza) [inline src/lib/io/include/sol-iio.h]: @yongli3 Please, increase version of SOL_IIO_CONFIG_API_VERSION.

**Comment by edersondisouza** [comment]
Sorry for the delay, but other than increasing struct version, PR looks good to me.

File: src/lib/io/include/sol-iio.h (hunk 4 above)
Comment (yongli3) [inline src/lib/io/include/sol-iio.h]: Thanks @edersondisouza, I increased the version to 2, please check if there is missing

**Comment by edersondisouza** [comment]
Merged, tks!

**Pull request closed without merging**